### PR TITLE
Add DATA_MATCH data consistency checker impl, support chunked streaming data consistency check, refactor data consitency check SPI definition and process

### PR DESCRIPTION
**Pull request by sandynz**
Fixes #13875 part 15.

Changes proposed in this pull request:
- Add DATA_MATCH data consistency checker impl
- Support chunked streaming data consistency check
- Support chunked streaming data calculation for DATA_MATCH
- Refactor data consistency check algorithm properties initialization
- Refactor SingleTableDataCalculator.getDatabaseTypes to support multiple databases
- Refactor CRC32_MATCH impl package; Refactor data consistency checker and calculator method signature; Reuse data source in data consistency check;
- Refactor DEFAULT DataConsistencyCheckAlgorithm to CRC32_MATCH
- Add calculate param for extension

### CODE REVIEW

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/config/datasource/ShardingSphereJDBCDataSourceCreator.java [hunk 1]
```diff
@@ -36,8 +38,19 @@
...
-                Collections.singletonList(ShardingRuleConfigurationConverter.findAndConvertShardingRuleConfiguration(rootConfig.getRules())), null);
+                Collections.singletonList(shardingRuleConfig), null);
+    }
+    
+    private void enableRangeQueryForInline(final ShardingRuleConfiguration shardingRuleConfig) {
+        for (ShardingSphereAlgorithmConfiguration each : shardingRuleConfig.getShardingAlgorithms().values()) {
+            if (!"INLINE".equalsIgnoreCase(each.getType())) {
+                continue;
+            }
+            each.getProps().put("allow-range-query-with-inline-sharding", "true");
```
Comment (taojintianxia) [inline shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/config/datasource/ShardingSphereJDBCDataSourceCreator.java:52]: maybe this should be a constant ?

**Review by menghaoranss — APPROVED** [review]

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java [hunk 2]
```diff
@@ -119,31 +126,49 @@ private long count(final DataSource dataSource, final String table, final Databa
                 jobContext.getJobConfig().getRuleConfig().getTarget().getType(), jobContext.getJobConfig().getRuleConfig().getTarget().getParameter());
         checkDatabaseTypeSupportedOrNot(supportedDatabaseTypes, targetConfig.getDatabaseType().getName());
         Collection<String> logicTableNames = jobContext.getTaskConfigs().stream().flatMap(each -> each.getDumperConfig().getTableNameMap().values().stream()).distinct().collect(Collectors.toList());
-        Map<String, Collection<String>> tablesColumnNamesMap = getTablesColumnNamesMap(sourceConfig);
+        Map<String, TableMetaData> tableMetaDataMap = getTablesColumnsMap(sourceConfig, logicTableNames);
         logicTableNames.forEach(each -> {
             //TODO put to preparer
-            if (!tablesColumnNamesMap.containsKey(each)) {
+            if (!tableMetaDataMap.containsKey(each)) {
                 throw new DataCheckFailException(String.format("could not get table columns for '%s'", each));
```
Comment (taojintianxia) [inline shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java:133]: not sure why upcon code like 
```
DataCheckFailException("count check failed, table=" + table, ex)
```

but here is String.format ?

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/config/datasource/ShardingSphereJDBCDataSourceCreator.java (hunk 1 above)
Comment (sandynz) [reply shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/config/datasource/ShardingSphereJDBCDataSourceCreator.java:52]: I tried to find the `allow-range-query-with-inline-sharding` constant definition, not defined in infra.

File: shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java (hunk 2 above)
Comment (sandynz) [reply shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java:133]: OK, I'll update the exception message style next time.

**Comment by sandynz** [comment]
Merged now, I'm working on next module refactoring PR.

**Pull request merged**